lib/pbench/server/api/resources/users_api.py

```diff
@@ -0,0 +1,533 @@
+                "data": return_data,
+            }
+            return make_response(jsonify(response_object), 200)
+        elif user.is_admin():
+            try:
+                target_user = User.query(username=username)
+                return_data = target_user.get_json()
+                response_object = {
+                    "status": "success",
+                    "data": return_data,
+                }
+                return make_response(jsonify(response_object), 200)
+            except Exception:
+                self.logger.exception(
+                    "Exception occurred while querying the user. Username: {}", username
+                )
+                abort(500, message="INTERNAL ERROR")
+        else:
+            self.logger.warning(
+                "Username retrieved from the auth token {} is different from the username provided",
+                self.auth.token_auth.current_user().id,
+            )
+            abort(403, message="Authentication token does not belong to the user")
+
+    @Auth.token_auth.login_required(f=Auth().verify_auth())
+    def put(self, username):
+        """
+        PUT request for updating user data.
+        This requires a Pbench auth token in the header field
+
+        This requires a JSON data with required user registration fields that needs an update
+        Example Json:
+        {
+            "first_name": "new_name",
+            "password": "password",
+            ...
+        }
+
+        Required headers include
+
+            Content-Type:   application/json
+            Accept:         application/json
+            Authorization:  Bearer Pbench_auth_token (user received upon login)
+
+        :return: JSON Payload
+            response_object = {
+                    "status": "success"/"fail",
+                    "message": "Successfully registered."/"failure message",
+                    "data": {
+                        "username": <username>,
+                        "first_name": <firstName>,
+                        "last_name": <lastName>,
+                        "registered_on": registered_on,
+                    }
+                }
+        """
+        post_data = request.get_json()
+        if not post_data:
+            self.logger.warning("Invalid json object: {}", request.url)
+            abort(400, message="Invalid json object in request")
+
+        try:
+            user, verified = self.auth.verify_user(username)
+        except Exception:
+            self.logger.exception("Exception occurred while verifying the user")
+            abort(500, message="INTERNAL ERROR")
+
+        # TODO: Check if the user has the right privileges
+        if verified:
+            # Check if the user payload contain fields that are either non-updatabale or
+            # are not present in the user db. If any key in the payload does not match
+            # with the column name we will abort the update request.
+            non_updatable = list(
```

Comment:
The use of "non-updatable" seems to imply that the caller specified fields in the `PUT` JSON body which are valid, but we just can't update.  However, but at line 447 we are reaching into the `User` class to get the list of all fields.  This means that a user could specify fields in the `PUT` body we don't allow.

Let's file an issue to fix this in a subsequent commit.

Reply:
Ah; the name should be `non-existent` instead of `non_updatable` I will fix that. We abort if there are any non-existent user column fields present in the payload. Whether the field is updatable or not we check that in user model.

Reply:
I think "the issue" is that we don't have a list of the fields which _can_ be modified, so we cannot generate the list of offending field references in the user's request.  (If we had a `User.__table__.user_modifiable_columns` attribute, we could use that in the expression at line 447, and then `non_updatable` would end up with the appropriate contents...although we still might want to compare it to `User.__table__.columns.keys()` in order to generate the appropriate diagnostics.)

Reply:
I didn't see @npalaska's comment before I made mine...so, nevermind.  :-)